presto-main/src/main/java/io/prestosql/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -346,19 +359,32 @@ else if (pendingSplits.isEmpty()) {
             return new ScheduleResult(false, overallNewTasks.build(), overallSplitAssignmentCount);
         }
 
-        if (anyBlockedOnPlacements || groupedExecution) {
-            // In a broadcast join, output buffers of the tasks in build source stage have to
-            // hold onto all data produced before probe side task scheduling finishes,
-            // even if the data is acknowledged by all known consumers. This is because
-            // new consumers may be added until the probe side task scheduling finishes.
-            //
-            // As a result, the following line is necessary to prevent deadlock
-            // due to neither build nor probe can make any progress.
-            // The build side blocks due to a full output buffer.
-            // In the meantime the probe side split cannot be consumed since
-            // builder side hash table construction has not finished.
+        // In a broadcast join, output buffers of the tasks in build source stage have to
+        // hold onto all data produced before probe side task scheduling finishes,
+        // even if the data is acknowledged by all known consumers. This is because
+        // new consumers may be added until the probe side task scheduling finishes.
+        //
+        // As a result, the following line is necessary to prevent deadlock
+        // due to neither build nor probe can make any progress.
+        // The build side blocks due to a full output buffer.
+        // In the meantime the probe side split cannot be consumed since
+        // builder side hash table construction has not finished.
+        if (groupedExecution) {
             overallNewTasks.addAll(finalizeTaskCreationIfNecessary());
         }
+        else if (anyBlockedOnPlacements) {
+            boolean shouldLockdownTasks = true;
+            if (sourceTasksProvider != null) {
```

Comment:
is this check still needed?

Reply:
the check is still needed because in FixedSourcePartitionedScheduler we set the supplier as null currently.